[BUG] Quick Install fails

**Describe the bug**
Installing via pre-built binaries or `go install` fails

**To Reproduce**
1A. `go install github.com/SamyRai/juleson/cmd/juleson@latest`
```
go/pkg/mod/github.com/!samy!rai/juleson@v0.0.0-20251209110105-2869b665de83/internal/cli/commands/dev.go:9:2: module github.com/SamyRai/juleson@latest found (v0.0.0-20251209110105-2869b665de83), but does not contain package github.com/SamyRai/juleson/internal/build
```

1B. `curl -L https://github.com/SamyRai/juleson/releases/latest/download/install.sh`
```
Not Found
```

**Expected behavior**
juleson installs

**Desktop (please complete the following information):**
 - OS: MacOS 26.2

**Additional context**
Add any other context about the problem here.

--- a/.github/workflows/ci.yml
+++ b/.github/workflows/ci.yml
@@ -257,6 +257,7 @@ jobs:
 
       - name: Test binaries (Windows)
         if: runner.os == 'Windows'
+        shell: pwsh
         run: |
           .\bin\juleson.exe --version || echo "Version flag not implemented yet"
           .\bin\jules-mcp.exe --help || echo "Help flag not implemented yet"
@@ -337,13 +338,28 @@ jobs:
 
     steps:
       - name: Check all job statuses
+        env:
+          PRE_CHECK_STATUS: ${{ needs.pre-check.result }}
+          TEST_STATUS: ${{ needs.test.result }}
+          LINT_STATUS: ${{ needs.lint.result }}
+          SECURITY_STATUS: ${{ needs.security.result }}
+          BUILD_STATUS: ${{ needs.build.result }}
+          DEPENDENCY_REVIEW_STATUS: ${{ needs.dependency-review.result }}
+          PERFORMANCE_TEST_STATUS: ${{ needs.performance-test.result }}
         run: |
-          # Define required jobs
-          required_jobs="pre-check test lint security build dependency-review performance-test"
+          declare -A job_statuses=(
+            ["pre-check"]="$PRE_CHECK_STATUS"
+            ["test"]="$TEST_STATUS"
+            ["lint"]="$LINT_STATUS"
+            ["security"]="$SECURITY_STATUS"
+            ["build"]="$BUILD_STATUS"
+            ["dependency-review"]="$DEPENDENCY_REVIEW_STATUS"
+            ["performance-test"]="$PERFORMANCE_TEST_STATUS"
+          )
 
           # Check each required job
-          for job in $required_jobs; do
-              status="${{ needs.$job.result }}"
+          for job in "${!job_statuses[@]}"; do
+              status="${job_statuses[$job]}"
               if [[ "$status" != "success" && "$status" != "skipped" ]]; then
                   echo "❌ Job '$job' failed with status: $status"
                   exit 1

--- a/.github/workflows/ci.yml
+++ b/.github/workflows/ci.yml
@@ -198,7 +198,7 @@ jobs:
           cache: true
 
       - name: Run Gosec Security Scanner
-        uses: securecodewarrior/github-action-gosec@master
+        uses: securego/gosec@v2.26.1
         with:
           args: "-no-fail -fmt sarif -out gosec-results.sarif ./..."
 
@@ -209,7 +209,7 @@ jobs:
           sarif_file: gosec-results.sarif
 
       - name: Run Trivy vulnerability scanner
-        uses: aquasecurity/trivy-action@master
+        uses: aquasecurity/trivy-action@v0.36.0
         with:
           scan-type: "fs"
           scan-ref: "."
@@ -291,7 +291,7 @@ jobs:
           fetch-depth: 1
 
       - name: Dependency Review
-        uses: actions/dependency-review-action@v4
+        uses: actions/dependency-review-action@v5
         with:
           fail-on-severity: moderate
           allow-licenses: MIT, Apache-2.0, BSD-2-Clause, BSD-3-Clause, ISC

--- a/.github/workflows/ci.yml
+++ b/.github/workflows/ci.yml
@@ -123,6 +123,8 @@ jobs:
         run: go mod download
 
       - name: Run tests with race detection
+        env:
+          CGO_ENABLED: 1
         run: go test -v -race -timeout=10m -coverprofile=coverage.out -covermode=atomic ./...
 
       - name: Run tests with -short flag

--- a/.github/workflows/ci.yml
+++ b/.github/workflows/ci.yml
@@ -125,7 +125,18 @@ jobs:
       - name: Run tests with race detection
         env:
           CGO_ENABLED: 1
-        run: go test -v -race -timeout=10m -coverprofile=coverage.out -covermode=atomic ./...
+          SKIP_E2E: 1
+        run: go test -v -race -timeout=10m ./...
+
+      - name: Run coverage
+        if: matrix.os == 'ubuntu-latest' && matrix.go-version == 'stable'
+        env:
+          SKIP_E2E: 1
+        run: go test -v -timeout=10m -coverprofile=coverage.out -covermode=atomic ./...
+
+      - name: Run MCP E2E
+        if: matrix.os == 'ubuntu-latest' && matrix.go-version == 'stable'
+        run: go test -v -run TestServerE2EWithCommandTransport ./internal/mcp
 
       - name: Run tests with -short flag
         run: go test -v -short -timeout=5m ./...
@@ -164,7 +175,7 @@ jobs:
       - name: Checkout code
         uses: actions/checkout@v6
         with:
-          fetch-depth: 1
+          fetch-depth: 0
 
       - name: Set up Go
         uses: actions/setup-go@v6
@@ -172,11 +183,21 @@ jobs:
           go-version-file: "go.mod"
           cache: true
 
+      - name: Determine lint base
+        id: lint-base
+        shell: bash
+        run: |
+          if [ "${{ github.event_name }}" = "pull_request" ]; then
+              echo "base=origin/${{ github.base_ref }}" >> "$GITHUB_OUTPUT"
+          else
+              echo "base=HEAD~1" >> "$GITHUB_OUTPUT"
+          fi
+
       - name: Run golangci-lint
         uses: golangci/golangci-lint-action@v9
         with:
           version: latest
-          args: --timeout=10m --verbose
+          args: --timeout=10m --verbose --new-from-rev=${{ steps.lint-base.outputs.base }}
           skip-cache: false
 
   security:
@@ -246,10 +267,32 @@ jobs:
           cache: true
 
       - name: Build juleson (optimized)
-        run: go build -v -ldflags="-s -w -X main.version=${{ github.sha }} -X main.buildTime=$(date -u +'%Y-%m-%dT%H:%M:%SZ')" -trimpath -o bin/juleson ./cmd/juleson
+        shell: bash
+        run: |
+          ext=""
+          if [ "$RUNNER_OS" = "Windows" ]; then
+              ext=".exe"
+          fi
+          build_time="$(date -u +'%Y-%m-%dT%H:%M:%SZ')"
+          go build -v \
+            -ldflags="-s -w -X main.version=${{ github.sha }} -X main.buildTime=${build_time}" \
+            -trimpath \
+            -o "bin/juleson${ext}" \
+            ./cmd/juleson
 
       - name: Build jules-mcp (optimized)
-        run: go build -v -ldflags="-s -w -X main.version=${{ github.sha }} -X main.buildTime=$(date -u +'%Y-%m-%dT%H:%M:%SZ')" -trimpath -o bin/jules-mcp ./cmd/jules-mcp
+        shell: bash
+        run: |
+          ext=""
+          if [ "$RUNNER_OS" = "Windows" ]; then
+              ext=".exe"
+          fi
+          build_time="$(date -u +'%Y-%m-%dT%H:%M:%SZ')"
+          go build -v \
+            -ldflags="-s -w -X main.version=${{ github.sha }} -X main.buildTime=${build_time}" \
+            -trimpath \
+            -o "bin/jules-mcp${ext}" \
+            ./cmd/jules-mcp
 
       - name: Test binaries (Unix)
         if: runner.os != 'Windows'
@@ -268,7 +311,7 @@ jobs:
         run: go vet ./...
 
       - name: Check for ineffective assignments
-        run: go run honnef.co/go/tools/cmd/staticcheck@latest ./...
+        run: go run honnef.co/go/tools/cmd/staticcheck@latest -checks=SA4006 ./...
 
       - name: Upload binaries
         if: matrix.os == 'ubuntu-latest'
@@ -319,7 +362,7 @@ jobs:
         run: go test -bench=. -benchmem -run=^$ ./...
 
       - name: Run memory profiling
-        run: go test -memprofile=mem.out -run=^$ ./...
+        run: go test -memprofile=mem.out -run=^$ ./internal/mcp
 
   all-checks-pass:
     name: All Checks Pass

--- a/.github/workflows/release.yml
+++ b/.github/workflows/release.yml
@@ -116,6 +116,8 @@ jobs:
             goarch: arm64
           - goos: windows
             goarch: amd64
+          - goos: windows
+            goarch: arm64
 
     steps:
       - name: Checkout code

--- a/.github/workflows/release.yml
+++ b/.github/workflows/release.yml
@@ -81,7 +81,8 @@ jobs:
         id: tag_check
         if: github.event_name == 'workflow_dispatch'
         run: |
-          if git show-ref --tags | grep -q "refs/tags/${{ github.event.inputs.version || github.ref_name }}"; then
+          VERSION="${{ github.event.inputs.version || github.ref_name }}"
+          if git show-ref --tags | grep -q "refs/tags/${VERSION}"; then
               echo "Tag already exists"
               exit 1
           fi
@@ -90,11 +91,11 @@ jobs:
         id: version
         run: |
           VERSION="${{ github.event.inputs.version || github.ref_name }}"
-          echo "version=$VERSION" >> $GITHUB_OUTPUT
+          echo "version=$VERSION" >> "$GITHUB_OUTPUT"
           if [[ $VERSION == *"beta"* || $VERSION == *"alpha"* || $VERSION == *"rc"* ]]; then
-              echo "is_prerelease=true" >> $GITHUB_OUTPUT
+              echo "is_prerelease=true" >> "$GITHUB_OUTPUT"
           else
-              echo "is_prerelease=false" >> $GITHUB_OUTPUT
+              echo "is_prerelease=false" >> "$GITHUB_OUTPUT"
           fi
 
   # Build release artifacts
@@ -245,12 +246,12 @@ jobs:
 
           # Generate changelog
           if command -v git-cliff &> /dev/null; then
-              git-cliff $PREVIOUS_TAG..HEAD --latest --strip header > changelog.md
+              git-cliff "$PREVIOUS_TAG..HEAD" --latest --strip header > changelog.md
           else
               # Fallback to git log
               echo "## Changes" > changelog.md
               echo "" >> changelog.md
-              git log --pretty=format:"- %s (%h)" $PREVIOUS_TAG..HEAD >> changelog.md
+              git log --pretty=format:"- %s (%h)" "$PREVIOUS_TAG..HEAD" >> changelog.md
           fi
 
       - name: Create Release

--- a/.github/workflows/release.yml
+++ b/.github/workflows/release.yml
@@ -79,6 +79,7 @@ jobs:
 
       - name: Check if tag exists
         id: tag_check
+        if: github.event_name == 'workflow_dispatch'
         run: |
           if git show-ref --tags | grep -q "refs/tags/${{ github.event.inputs.version || github.ref_name }}"; then
               echo "Tag already exists"
@@ -149,6 +150,23 @@ jobs:
             -o dist/jules-mcp-${{ matrix.goos }}-${{ matrix.goarch }}${EXT} \
             ./cmd/jules-mcp
 
+      - name: Package release assets
+        run: |
+          set -euo pipefail
+          OS="${{ matrix.goos }}"
+          ARCH="${{ matrix.goarch }}"
+
+          if [ "$OS" = "windows" ]; then
+              (cd dist && zip -q "juleson-${OS}-${ARCH}.zip" "juleson-${OS}-${ARCH}.exe")
+              (cd dist && zip -q "jules-mcp-${OS}-${ARCH}.zip" "jules-mcp-${OS}-${ARCH}.exe")
+          else
+              mkdir -p package/juleson package/jules-mcp
+              cp "dist/juleson-${OS}-${ARCH}" package/juleson/juleson
+              cp "dist/jules-mcp-${OS}-${ARCH}" package/jules-mcp/jules-mcp
+              tar -C package/juleson -czf "dist/juleson-${OS}-${ARCH}.tar.gz" juleson
+              tar -C package/jules-mcp -czf "dist/jules-mcp-${OS}-${ARCH}.tar.gz" jules-mcp
+          fi
+
       - name: Upload build artifacts
         uses: actions/upload-artifact@v7
         with:
@@ -178,20 +196,21 @@ jobs:
       - name: Publish to Go Module Proxy
         run: |
           # Set GOPROXY to ensure we publish to the official proxy
-          GOPROXY=proxy.golang.org go list -m juleson@${{ needs.validate.outputs.version }}
+          GOPROXY=proxy.golang.org go list -m github.com/SamyRai/juleson@${{ needs.validate.outputs.version }}
 
       - name: Verify module is published
         run: |
           # Wait a moment for indexing
           sleep 30
           # Verify the module is available
-          go list -m -versions juleson | grep "${{ needs.validate.outputs.version }}"
+          go list -m -versions github.com/SamyRai/juleson | grep "${{ needs.validate.outputs.version }}"
 
   # Create GitHub release
   release:
     name: Create GitHub Release
     needs: [validate, build, publish-module]
     runs-on: ubuntu-latest
+    if: always() && needs.validate.result == 'success' && needs.build.result == 'success' && (needs.publish-module.result == 'success' || needs.publish-module.result == 'skipped')
 
     steps:
       - name: Checkout code
@@ -203,6 +222,11 @@ jobs:
         uses: actions/download-artifact@v8
         with:
           path: dist/
+
+      - name: Add installer scripts
+        run: |
+          cp scripts/install.sh dist/install.sh
+          cp scripts/install.ps1 dist/install.ps1
 
       - name: Generate checksums
         run: |

--- a/CONTRIBUTING.md
+++ b/CONTRIBUTING.md
@@ -11,7 +11,7 @@ By participating in this project, you agree to maintain a respectful and inclusi
 
 ### Prerequisites
 
-- Go 1.23 or higher
+- Go 1.25 or higher
 - Git
 - Jules API access (for integration testing)
 

--- a/README.md
+++ b/README.md
@@ -259,7 +259,7 @@ Juleson/
 
 ### **Prerequisites**
 
-- Go 1.24 or higher
+- Go 1.25 or higher
 - Jules API key ([Get one from Google](https://jules.googleapis.com))
 - Git (for project analysis features)
 - Optional: Gemini API key (for AI orchestration features)
@@ -274,17 +274,23 @@ Juleson/
 **Linux/macOS:**
 
 ```bash
-# Using Go (requires Go 1.24+)
+# Using pre-built binaries
+curl -L https://github.com/SamyRai/juleson/releases/latest/download/install.sh | bash
+
+# Or using Go (requires Go 1.25+)
 go install github.com/SamyRai/juleson/cmd/juleson@latest
-go install github.com/SamyRai/juleson/cmd/juleson-mcp@latest
+go install github.com/SamyRai/juleson/cmd/jules-mcp@latest
 ```
 
 **Windows:**
 
 ```powershell
-# Using Go (requires Go 1.24+)
+# Using pre-built binaries
+irm https://github.com/SamyRai/juleson/releases/latest/download/install.ps1 | iex
+
+# Or using Go (requires Go 1.25+)
 go install github.com/SamyRai/juleson/cmd/juleson@latest
-go install github.com/SamyRai/juleson/cmd/juleson-mcp@latest
+go install github.com/SamyRai/juleson/cmd/jules-mcp@latest
 ```
 
 #### Build from Source

--- a/README.md
+++ b/README.md
@@ -32,7 +32,7 @@ Juleson bridges your development workflow with Google's Jules AI agent, providin
 Juleson/
 ├── cmd/                          # Application entry points
 │   ├── juleson/                 # CLI tool for direct usage
-│   ├── juleson-mcp/             # MCP server for AI assistants
+│   ├── jules-mcp/               # MCP server for AI assistants
 │   └── orchestrator/            # Build orchestrator
 ├── internal/
 │   ├── agent/                   # 🤖 Intelligent AI Agent System
@@ -321,7 +321,7 @@ go build -o bin/orchestrator ./cmd/orchestrator
 
 # Verify installation
 juleson --version
-juleson-mcp --version
+jules-mcp --version
 ```
 
 ## 📖 **Usage**
@@ -396,7 +396,7 @@ juleson template create my-template refactoring "Description" # Create custom te
 Start the MCP server for integration with AI assistants:
 
 ```bash
-./bin/juleson-mcp
+./bin/jules-mcp
 ```
 
 #### **Configure with Claude Desktop**
@@ -407,7 +407,7 @@ Add to `~/Library/Application Support/Claude/claude_desktop_config.json`:
 {
   "mcpServers": {
     "Juleson": {
-      "command": "/absolute/path/to/Juleson/bin/juleson-mcp",
+      "command": "/absolute/path/to/Juleson/bin/jules-mcp",
       "env": {
         "JULES_API_KEY": "your-api-key"
       }
@@ -424,7 +424,7 @@ Add to Cursor settings JSON:
 {
   "mcp.servers": {
     "Juleson": {
-      "command": "/absolute/path/to/Juleson/bin/juleson-mcp",
+      "command": "/absolute/path/to/Juleson/bin/jules-mcp",
       "env": {
         "JULES_API_KEY": "your-api-key"
       }

--- a/TESTING_GUIDE.md
+++ b/TESTING_GUIDE.md
@@ -120,7 +120,7 @@ binary/script boundary
 
 Current E2E coverage includes:
 
-- `internal/mcp/server_e2e_test.go`, which builds a temporary `juleson-mcp`
+- `internal/mcp/server_e2e_test.go`, which builds a temporary `jules-mcp`
   binary when `JULESON_MCP_BINARY` is not set, connects through MCP command
   transport, and verifies tools, prompts, and resources.
 - `scripts/install_test.go`, which serves local fake release archives and runs

--- a/TESTING_GUIDE.md
+++ b/TESTING_GUIDE.md
@@ -115,7 +115,19 @@ func TestClient_SessionWorkflow(t *testing.T) {
 
 **Purpose**: Test complete user workflows
 
-**Location**: `e2e/` directory or `*_e2e_test.go`
+**Location**: `*_e2e_test.go` files or package-level tests that execute the real
+binary/script boundary
+
+Current E2E coverage includes:
+
+- `internal/mcp/server_e2e_test.go`, which builds a temporary `juleson-mcp`
+  binary when `JULESON_MCP_BINARY` is not set, connects through MCP command
+  transport, and verifies tools, prompts, and resources.
+- `scripts/install_test.go`, which serves local fake release archives and runs
+  the real `scripts/install.sh` installer against a temporary install directory.
+
+E2E tests are skipped by `go test -short ./...`. Set `SKIP_E2E=1` to skip MCP
+E2E in regular test runs when debugging locally.
 
 **Example**:
 
@@ -263,6 +275,22 @@ go test -tags=integration ./...
 
 # Run with real API (requires API key)
 JULES_API_KEY=your-key go test -tags=integration ./internal/jules/
+```
+
+### End-to-End Tests
+
+```bash
+# Run the full unit, integration-style, and E2E suite
+go test ./...
+
+# Run only MCP command-transport E2E
+go test -run TestServerE2EWithCommandTransport ./internal/mcp
+
+# Run installer unit and E2E checks
+go test ./scripts
+
+# Skip E2E during a short pass
+go test -short ./...
 ```
 
 ## Writing Tests

--- a/docs/DEPLOYMENT_GUIDE.md
+++ b/docs/DEPLOYMENT_GUIDE.md
@@ -163,11 +163,11 @@ brew install Juleson
 
 ```bash
 # Install latest version
-go install github.com/SamyRai/Juleson/cmd/juleson@latest
-go install github.com/SamyRai/Juleson/cmd/jules-mcp@latest
+go install github.com/SamyRai/juleson/cmd/juleson@latest
+go install github.com/SamyRai/juleson/cmd/jules-mcp@latest
 
 # Install specific version
-go install github.com/SamyRai/Juleson/cmd/juleson@v0.1.0
+go install github.com/SamyRai/juleson/cmd/juleson@v0.1.0
 ```
 
 ---

--- a/docs/INSTALLATION_GUIDE.md
+++ b/docs/INSTALLATION_GUIDE.md
@@ -115,7 +115,7 @@ sudo mv juleson jules-mcp /usr/local/bin/
 sudo chmod +x /usr/local/bin/juleson /usr/local/bin/jules-mcp
 ```
 
-### Windows (x86_64)
+### Windows (x86_64 and ARM64)
 
 ```powershell
 # Download binaries
@@ -137,6 +137,9 @@ Expand-Archive -Path "jules-mcp.zip" -DestinationPath "C:\Program Files\juleson"
 juleson --version
 jules-mcp --version
 ```
+
+On Windows ARM64, use `juleson-windows-arm64.zip` and
+`jules-mcp-windows-arm64.zip`.
 
 ## Method 2: Go Install
 

--- a/docs/INSTALLATION_GUIDE.md
+++ b/docs/INSTALLATION_GUIDE.md
@@ -22,7 +22,7 @@ method that best fits your needs.
 # Using pre-built binaries
 curl -L https://github.com/SamyRai/juleson/releases/latest/download/install.sh | bash
 
-# Or using Go (requires Go 1.23+)
+# Or using Go (requires Go 1.25+)
 go install github.com/SamyRai/juleson/cmd/juleson@latest
 go install github.com/SamyRai/juleson/cmd/jules-mcp@latest
 ```
@@ -33,7 +33,7 @@ go install github.com/SamyRai/juleson/cmd/jules-mcp@latest
 # Using PowerShell
 irm https://github.com/SamyRai/juleson/releases/latest/download/install.ps1 | iex
 
-# Or using Go (requires Go 1.23+)
+# Or using Go (requires Go 1.25+)
 go install github.com/SamyRai/juleson/cmd/juleson@latest
 go install github.com/SamyRai/juleson/cmd/jules-mcp@latest
 ```
@@ -140,7 +140,7 @@ jules-mcp --version
 
 ## Method 2: Go Install
 
-Requires Go 1.23 or higher.
+Requires Go 1.25 or higher.
 
 ### All Platforms
 
@@ -184,7 +184,7 @@ $goPath = go env GOPATH
 
 ### Prerequisites
 
-- Go 1.23 or higher
+- Go 1.25 or higher
 - Git
 - Make (Linux/macOS) or PowerShell (Windows)
 

--- a/docs/README.md
+++ b/docs/README.md
@@ -71,7 +71,7 @@ See [`templates/`](../templates/README.md) for template documentation.
 Juleson/
 ├── cmd/                          # Application entry points
 │   ├── juleson/                 # CLI tool
-│   └── juleson-mcp/             # MCP server
+│   └── jules-mcp/               # MCP server
 ├── internal/                     # Core packages
 │   ├── events/                  # Event-driven architecture
 │   ├── jules/                   # Jules API client

--- a/internal/agent/core/agent_test.go
+++ b/internal/agent/core/agent_test.go
@@ -26,8 +26,7 @@ func TestExecuteSingleTask(t *testing.T) {
 		EnableTelemetry: false,
 	}
 
-	// Create agent
-	coreAgent := NewAgent(registry, config).(*CoreAgent)
+	coreAgent := newTestCoreAgent(t, registry, config)
 
 	// Create a test task
 	task := &agent.Task{
@@ -77,7 +76,7 @@ func TestPrepareToolParameters(t *testing.T) {
 
 	// Create agent
 	registry := tools.NewToolRegistry()
-	coreAgent := NewAgent(registry, config).(*CoreAgent)
+	coreAgent := newTestCoreAgent(t, registry, config)
 
 	// Create a test task
 	task := &agent.Task{
@@ -174,4 +173,13 @@ func TestNewAgent(t *testing.T) {
 	if coreAgent.maxIterations != DefaultMaxIterations {
 		t.Errorf("Expected maxIterations %d, got %d", DefaultMaxIterations, coreAgent.maxIterations)
 	}
+}
+
+func newTestCoreAgent(t *testing.T, registry tools.ToolRegistry, config *Config) *CoreAgent {
+	t.Helper()
+	coreAgent, ok := NewAgent(registry, config).(*CoreAgent)
+	if !ok {
+		t.Fatal("Expected CoreAgent type")
+	}
+	return coreAgent
 }

--- a/internal/agent/tools/registry_test.go
+++ b/internal/agent/tools/registry_test.go
@@ -90,7 +90,7 @@ func TestGet(t *testing.T) {
 	registry := NewToolRegistry()
 
 	tool := &mockTool{name: "test-tool"}
-	registry.Register(tool)
+	mustRegister(t, registry, tool)
 
 	// Get existing tool
 	retrieved, err := registry.Get("test-tool")
@@ -115,9 +115,9 @@ func TestList(t *testing.T) {
 	tool2 := &mockTool{name: "tool2"}
 	tool3 := &mockTool{name: "tool3"}
 
-	registry.Register(tool1)
-	registry.Register(tool2)
-	registry.Register(tool3)
+	mustRegister(t, registry, tool1)
+	mustRegister(t, registry, tool2)
+	mustRegister(t, registry, tool3)
 
 	tools := registry.List()
 	if len(tools) != 3 {
@@ -158,9 +158,9 @@ func TestFindForTask(t *testing.T) {
 		},
 	}
 
-	registry.Register(tool1)
-	registry.Register(tool2)
-	registry.Register(toolAll)
+	mustRegister(t, registry, tool1)
+	mustRegister(t, registry, tool2)
+	mustRegister(t, registry, toolAll)
 
 	// Test finding tools for task1
 	task1 := agent.Task{Name: "task1"}
@@ -195,7 +195,9 @@ func TestConcurrentAccess(t *testing.T) {
 	for i := 0; i < 10; i++ {
 		go func(n int) {
 			tool := &mockTool{name: fmt.Sprintf("tool%d", n)}
-			registry.Register(tool)
+			if err := registry.Register(tool); err != nil {
+				t.Errorf("register tool: %v", err)
+			}
 			done <- true
 		}(i)
 	}
@@ -207,5 +209,12 @@ func TestConcurrentAccess(t *testing.T) {
 	tools := registry.List()
 	if len(tools) != 10 {
 		t.Errorf("expected 10 tools after concurrent registration, got %d", len(tools))
+	}
+}
+
+func mustRegister(t *testing.T, registry ToolRegistry, tool Tool) {
+	t.Helper()
+	if err := registry.Register(tool); err != nil {
+		t.Fatalf("register %q: %v", tool.Name(), err)
 	}
 }

--- a/internal/cli/commands/dev.go
+++ b/internal/cli/commands/dev.go
@@ -39,7 +39,7 @@ func buildBinaries(ctx context.Context, version, goos, goarch string, race bool)
 		path string
 	}{
 		{"juleson", "./cmd/juleson"},
-		{"juleson-mcp", "./cmd/jules-mcp"},
+		{"jules-mcp", "./cmd/jules-mcp"},
 	}
 
 	for _, binary := range binaries {
@@ -574,7 +574,7 @@ func newInstallCommand() *cobra.Command {
 			}
 
 			// Install binaries
-			installer := build.NewInstaller("bin", []string{"juleson", "juleson-mcp"})
+			installer := build.NewInstaller("bin", []string{"juleson", "jules-mcp"})
 
 			var result *build.InstallResult
 

--- a/internal/github/actions.go
+++ b/internal/github/actions.go
@@ -101,8 +101,9 @@ func (s *ActionsService) ListWorkflowRuns(ctx context.Context, owner, repo strin
 		runs, _, err = s.client.Client.Actions.ListRepositoryWorkflowRuns(ctx, owner, repo, opts)
 	} else {
 		// Try as filename first
-		runs, _, err = s.client.Client.Actions.ListWorkflowRunsByFileName(ctx, owner, repo, workflowIDOrFile, opts)
-		if err != nil {
+		if runsByFile, _, fileErr := s.client.Client.Actions.ListWorkflowRunsByFileName(ctx, owner, repo, workflowIDOrFile, opts); fileErr == nil {
+			runs = runsByFile
+		} else {
 			// Try as ID
 			workflowID := int64(parseInt(workflowIDOrFile))
 			runs, _, err = s.client.Client.Actions.ListWorkflowRunsByID(ctx, owner, repo, workflowID, opts)
@@ -110,6 +111,9 @@ func (s *ActionsService) ListWorkflowRuns(ctx context.Context, owner, repo strin
 				return nil, fmt.Errorf("failed to list workflow runs: %w", err)
 			}
 		}
+	}
+	if err != nil {
+		return nil, fmt.Errorf("failed to list workflow runs: %w", err)
 	}
 
 	var result []*WorkflowRun

--- a/internal/jules/patches_test.go
+++ b/internal/jules/patches_test.go
@@ -5,9 +5,11 @@ import (
 	"os"
 	"path/filepath"
 	"testing"
+	"time"
 
 	"github.com/jarcoal/httpmock"
 	"github.com/stretchr/testify/assert"
+	"github.com/stretchr/testify/require"
 	"github.com/stretchr/testify/suite"
 )
 
@@ -18,7 +20,7 @@ type PatchesTestSuite struct {
 
 func (suite *PatchesTestSuite) SetupTest() {
 	httpmock.Activate()
-	suite.client = NewClient("test-api-key", "https://jules.googleapis.com", 30, 3)
+	suite.client = NewClient("test-api-key", "https://jules.googleapis.com", 30*time.Second, 3)
 }
 
 func (suite *PatchesTestSuite) TearDownTest() {
@@ -67,8 +69,8 @@ index 1234567..abcdefg 100644
 
 	changes, err := suite.client.GetSessionChanges(context.Background(), sessionID)
 
-	assert.NoError(suite.T(), err)
-	assert.NotNil(suite.T(), changes)
+	require.NoError(suite.T(), err)
+	require.NotNil(suite.T(), changes)
 	assert.Equal(suite.T(), sessionID, changes.SessionID)
 	assert.Equal(suite.T(), 1, changes.TotalPatches)
 	assert.Len(suite.T(), changes.Files, 1)

--- a/internal/mcp/server_e2e_test.go
+++ b/internal/mcp/server_e2e_test.go
@@ -4,6 +4,7 @@ import (
 	"context"
 	"os"
 	"os/exec"
+	"path/filepath"
 	"testing"
 	"time"
 
@@ -13,31 +14,25 @@ import (
 )
 
 func TestServerE2EWithCommandTransport(t *testing.T) {
-	// Use the binary from go/bin or current directory
-	binaryPath := os.Getenv("JULESON_MCP_BINARY")
-	if binaryPath == "" {
-		// Try common locations
-		for _, path := range []string{
-			"../bin/juleson-mcp",
-			"./bin/juleson-mcp",
-			"juleson-mcp",
-		} {
-			if _, err := os.Stat(path); err == nil {
-				binaryPath = path
-				break
-			}
-		}
+	if testing.Short() {
+		t.Skip("skipping command-transport E2E test in short mode")
 	}
-	if binaryPath == "" || os.Getenv("SKIP_E2E") != "" {
-		t.Skip("juleson-mcp binary not found or E2E tests disabled, run 'make build-mcp' first")
+	if os.Getenv("SKIP_E2E") != "" {
+		t.Skip("E2E tests disabled by SKIP_E2E")
 	}
 
 	ctx, cancel := context.WithTimeout(context.Background(), 10*time.Second)
 	defer cancel()
 
+	binaryPath := resolveMCPBinary(t, ctx)
+
 	// Create a command transport like VSCode would
 	cmd := exec.Command(binaryPath)
-	// Don't change directory - run from current directory
+	cmd.Dir = repoRoot(t)
+	cmd.Env = append(os.Environ(),
+		"HOME="+t.TempDir(),
+		"JULES_API_KEY=",
+	)
 	transport := &mcp.CommandTransport{Command: cmd}
 
 	// Create client
@@ -88,4 +83,56 @@ func TestServerE2EWithCommandTransport(t *testing.T) {
 			t.Logf("Found resource: %s - %s", resource.Name, resource.Description)
 		}
 	})
+}
+
+func resolveMCPBinary(t *testing.T, ctx context.Context) string {
+	t.Helper()
+
+	if binaryPath := os.Getenv("JULESON_MCP_BINARY"); binaryPath != "" {
+		if _, err := os.Stat(binaryPath); err != nil {
+			t.Fatalf("JULESON_MCP_BINARY %q is not usable: %v", binaryPath, err)
+		}
+		return binaryPath
+	}
+
+	root := repoRoot(t)
+	for _, path := range []string{
+		filepath.Join(root, "bin", "juleson-mcp"),
+		filepath.Join(root, "juleson-mcp"),
+	} {
+		if _, err := os.Stat(path); err == nil {
+			return path
+		}
+	}
+
+	binaryPath := filepath.Join(t.TempDir(), "juleson-mcp")
+	if os.Getenv("GOOS") == "windows" {
+		binaryPath += ".exe"
+	}
+
+	buildCmd := exec.CommandContext(ctx, "go", "build", "-o", binaryPath, "./cmd/jules-mcp")
+	buildCmd.Dir = root
+	if output, err := buildCmd.CombinedOutput(); err != nil {
+		t.Fatalf("build juleson-mcp E2E binary: %v\n%s", err, output)
+	}
+	return binaryPath
+}
+
+func repoRoot(t *testing.T) string {
+	t.Helper()
+
+	dir, err := os.Getwd()
+	if err != nil {
+		t.Fatal(err)
+	}
+	for {
+		if _, err := os.Stat(filepath.Join(dir, "go.mod")); err == nil {
+			return dir
+		}
+		parent := filepath.Dir(dir)
+		if parent == dir {
+			t.Fatal("could not find repository root")
+		}
+		dir = parent
+	}
 }

--- a/internal/mcp/server_e2e_test.go
+++ b/internal/mcp/server_e2e_test.go
@@ -97,15 +97,15 @@ func resolveMCPBinary(t *testing.T, ctx context.Context) string {
 
 	root := repoRoot(t)
 	for _, path := range []string{
-		filepath.Join(root, "bin", "juleson-mcp"),
-		filepath.Join(root, "juleson-mcp"),
+		filepath.Join(root, "bin", "jules-mcp"),
+		filepath.Join(root, "jules-mcp"),
 	} {
 		if _, err := os.Stat(path); err == nil {
 			return path
 		}
 	}
 
-	binaryPath := filepath.Join(t.TempDir(), "juleson-mcp")
+	binaryPath := filepath.Join(t.TempDir(), "jules-mcp")
 	if os.Getenv("GOOS") == "windows" {
 		binaryPath += ".exe"
 	}
@@ -113,7 +113,7 @@ func resolveMCPBinary(t *testing.T, ctx context.Context) string {
 	buildCmd := exec.CommandContext(ctx, "go", "build", "-o", binaryPath, "./cmd/jules-mcp")
 	buildCmd.Dir = root
 	if output, err := buildCmd.CombinedOutput(); err != nil {
-		t.Fatalf("build juleson-mcp E2E binary: %v\n%s", err, output)
+		t.Fatalf("build jules-mcp E2E binary: %v\n%s", err, output)
 	}
 	return binaryPath
 }

--- a/internal/mcp/server_e2e_test.go
+++ b/internal/mcp/server_e2e_test.go
@@ -5,6 +5,7 @@ import (
 	"os"
 	"os/exec"
 	"path/filepath"
+	"runtime"
 	"testing"
 	"time"
 
@@ -21,7 +22,7 @@ func TestServerE2EWithCommandTransport(t *testing.T) {
 		t.Skip("E2E tests disabled by SKIP_E2E")
 	}
 
-	ctx, cancel := context.WithTimeout(context.Background(), 10*time.Second)
+	ctx, cancel := context.WithTimeout(context.Background(), 30*time.Second)
 	defer cancel()
 
 	binaryPath := resolveMCPBinary(t, ctx)
@@ -89,6 +90,7 @@ func resolveMCPBinary(t *testing.T, ctx context.Context) string {
 	t.Helper()
 
 	if binaryPath := os.Getenv("JULESON_MCP_BINARY"); binaryPath != "" {
+		// #nosec G304,G703 -- tests intentionally accept an explicit local binary path.
 		if _, err := os.Stat(binaryPath); err != nil {
 			t.Fatalf("JULESON_MCP_BINARY %q is not usable: %v", binaryPath, err)
 		}
@@ -106,7 +108,7 @@ func resolveMCPBinary(t *testing.T, ctx context.Context) string {
 	}
 
 	binaryPath := filepath.Join(t.TempDir(), "jules-mcp")
-	if os.Getenv("GOOS") == "windows" {
+	if runtime.GOOS == "windows" {
 		binaryPath += ".exe"
 	}
 

--- a/internal/mcp/tools/dev.go
+++ b/internal/mcp/tools/dev.go
@@ -213,7 +213,7 @@ func buildProjectHandler(ctx context.Context, req *mcp.CallToolRequest, input Bu
 	// Build MCP
 	if buildMCP {
 		totalCount++
-		config := build.DefaultConfig("juleson-mcp", "./cmd/jules-mcp")
+		config := build.DefaultConfig("jules-mcp", "./cmd/jules-mcp")
 		config.Version = input.Version
 		config.GOOS = input.GOOS
 		config.GOARCH = input.GOARCH
@@ -533,7 +533,7 @@ func buildReleaseHandler(ctx context.Context, req *mcp.CallToolRequest, input Bu
 			path string
 		}{
 			{"juleson", "./cmd/juleson"},
-			{"juleson-mcp", "./cmd/jules-mcp"},
+			{"jules-mcp", "./cmd/jules-mcp"},
 		} {
 			totalCount++
 

--- a/internal/mcp/tools/github.go
+++ b/internal/mcp/tools/github.go
@@ -140,22 +140,30 @@ func createSessionFromRepo(ctx context.Context, req *mcp.CallToolRequest, input 
 
 		// Create session with explicit repo
 		session, err = ghClient.Sessions.CreateSessionFromRepo(ctx, input.Prompt, owner, name, input.Branch)
+		if err != nil {
+			return &mcp.CallToolResult{
+				IsError: true,
+				Content: []mcp.Content{
+					&mcp.TextContent{Text: fmt.Sprintf("Failed to create session: %v", err)},
+				},
+			}, GitHubCreateSessionOutput{}, err
+		}
 	} else {
 		// Auto-detect from current directory
 		session, err = ghClient.Sessions.CreateSessionFromCurrentRepo(ctx, input.Prompt, input.Branch)
-		if err == nil {
-			// Get repo info for output
-			repo, _ = ghClient.Repositories.DiscoverCurrentRepo(ctx)
+		if err != nil {
+			return &mcp.CallToolResult{
+				IsError: true,
+				Content: []mcp.Content{
+					&mcp.TextContent{Text: fmt.Sprintf("Failed to create session: %v", err)},
+				},
+			}, GitHubCreateSessionOutput{}, err
 		}
-	}
-
-	if err != nil {
-		return &mcp.CallToolResult{
-			IsError: true,
-			Content: []mcp.Content{
-				&mcp.TextContent{Text: fmt.Sprintf("Failed to create session: %v", err)},
-			},
-		}, GitHubCreateSessionOutput{}, err
+		// Get repo info for output when discovery succeeds.
+		discoveredRepo, discoverErr := ghClient.Repositories.DiscoverCurrentRepo(ctx)
+		if discoverErr == nil {
+			repo = discoveredRepo
+		}
 	}
 
 	branch := input.Branch

--- a/internal/orchestrator/orchestrator.go
+++ b/internal/orchestrator/orchestrator.go
@@ -28,7 +28,7 @@ type Config struct {
 func DefaultConfig(version, buildDate, gitCommit string) *Config {
 	return &Config{
 		BinaryCLI:    "juleson",
-		BinaryMCP:    "juleson-mcp",
+		BinaryMCP:    "jules-mcp",
 		BinDir:       "bin",
 		CmdCLIDir:    "cmd/juleson",
 		CmdMCPDir:    "cmd/jules-mcp",

--- a/scripts/README.md
+++ b/scripts/README.md
@@ -2,29 +2,27 @@
 
 This directory contains utility and automation scripts.
 
-## Planned Scripts
+## Installer Scripts
 
-- `build.sh` - Build script for all binaries
-- `test.sh` - Run all tests
-- `install.sh` - Installation script
-- `deploy.sh` - Deployment automation
-- `release.sh` - Release preparation
+- `install.sh` - Linux/macOS release installer for `juleson` and `jules-mcp`
+- `install.ps1` - Windows release installer for `juleson` and `jules-mcp`
 
 ## Usage
 
-Scripts should be executable and documented with comments explaining their purpose and usage.
-
-### Example: build.sh
+### Linux/macOS
 
 ```bash
-#!/bin/bash
-# Build all binaries
+curl -L https://github.com/SamyRai/juleson/releases/latest/download/install.sh | bash
+```
 
-echo "Building juleson..."
-go build -o bin/juleson ./cmd/juleson
+Use `INSTALL_DIR` to install somewhere other than `/usr/local/bin`:
 
-echo "Building jules-mcp..."
-go build -o bin/jules-mcp ./cmd/jules-mcp
+```bash
+INSTALL_DIR="$HOME/.local/bin" bash install.sh
+```
 
-echo "Build complete!"
+### Windows
+
+```powershell
+irm https://github.com/SamyRai/juleson/releases/latest/download/install.ps1 | iex
 ```

--- a/scripts/install.ps1
+++ b/scripts/install.ps1
@@ -2,6 +2,7 @@ param(
     [string]$Version = "latest",
     [string]$InstallDir = "$env:USERPROFILE\.juleson\bin",
     [string]$Repo = "SamyRai/juleson",
+    [string]$BaseUrl = $env:JULESON_INSTALL_BASE_URL,
     [switch]$NoPathUpdate,
     [switch]$Help
 )
@@ -14,7 +15,7 @@ function Show-Usage {
 Install the latest Juleson release binaries.
 
 Usage:
-  .\install.ps1 [-Version <tag|latest>] [-InstallDir <path>] [-Repo <owner/repo>] [-NoPathUpdate]
+  .\install.ps1 [-Version <tag|latest>] [-InstallDir <path>] [-Repo <owner/repo>] [-BaseUrl <url>] [-NoPathUpdate]
 
 Examples:
   irm https://github.com/SamyRai/juleson/releases/latest/download/install.ps1 | iex
@@ -37,7 +38,9 @@ $arch = switch ($env:PROCESSOR_ARCHITECTURE) {
     default { throw "Unsupported Windows architecture: $env:PROCESSOR_ARCHITECTURE" }
 }
 
-if ($Version -eq "latest") {
+if (-not [string]::IsNullOrWhiteSpace($BaseUrl)) {
+    $baseUrl = $BaseUrl.TrimEnd("/")
+} elseif ($Version -eq "latest") {
     $baseUrl = "https://github.com/$Repo/releases/latest/download"
 } else {
     $baseUrl = "https://github.com/$Repo/releases/download/$Version"

--- a/scripts/install.ps1
+++ b/scripts/install.ps1
@@ -32,9 +32,14 @@ if ($Help) {
 if ([string]::IsNullOrWhiteSpace($Version) -or [string]::IsNullOrWhiteSpace($InstallDir) -or [string]::IsNullOrWhiteSpace($Repo)) {
     throw "Version, install directory, and repository must be non-empty."
 }
+$installRoot = [System.IO.Path]::GetPathRoot($InstallDir)
+if ($InstallDir -ne $installRoot) {
+    $InstallDir = $InstallDir.TrimEnd('\', '/')
+}
 
 $arch = switch ($env:PROCESSOR_ARCHITECTURE) {
     "AMD64" { "amd64" }
+    "ARM64" { "arm64" }
     default { throw "Unsupported Windows architecture: $env:PROCESSOR_ARCHITECTURE" }
 }
 

--- a/scripts/install.ps1
+++ b/scripts/install.ps1
@@ -1,0 +1,82 @@
+param(
+    [string]$Version = "latest",
+    [string]$InstallDir = "$env:USERPROFILE\.juleson\bin",
+    [string]$Repo = "SamyRai/juleson",
+    [switch]$NoPathUpdate,
+    [switch]$Help
+)
+
+Set-StrictMode -Version Latest
+$ErrorActionPreference = "Stop"
+
+function Show-Usage {
+    @"
+Install the latest Juleson release binaries.
+
+Usage:
+  .\install.ps1 [-Version <tag|latest>] [-InstallDir <path>] [-Repo <owner/repo>] [-NoPathUpdate]
+
+Examples:
+  irm https://github.com/SamyRai/juleson/releases/latest/download/install.ps1 | iex
+  .\install.ps1 -InstallDir "$env:USERPROFILE\bin"
+  .\install.ps1 -Version v1.0.0 -NoPathUpdate
+"@
+}
+
+if ($Help) {
+    Show-Usage
+    exit 0
+}
+
+if ([string]::IsNullOrWhiteSpace($Version) -or [string]::IsNullOrWhiteSpace($InstallDir) -or [string]::IsNullOrWhiteSpace($Repo)) {
+    throw "Version, install directory, and repository must be non-empty."
+}
+
+$arch = switch ($env:PROCESSOR_ARCHITECTURE) {
+    "AMD64" { "amd64" }
+    default { throw "Unsupported Windows architecture: $env:PROCESSOR_ARCHITECTURE" }
+}
+
+if ($Version -eq "latest") {
+    $baseUrl = "https://github.com/$Repo/releases/latest/download"
+} else {
+    $baseUrl = "https://github.com/$Repo/releases/download/$Version"
+}
+
+$tempDir = Join-Path ([System.IO.Path]::GetTempPath()) ("juleson-install-" + [System.Guid]::NewGuid().ToString("N"))
+New-Item -ItemType Directory -Path $tempDir | Out-Null
+
+try {
+    New-Item -ItemType Directory -Force -Path $InstallDir | Out-Null
+
+    foreach ($binary in @("juleson", "jules-mcp")) {
+        $asset = "$binary-windows-$arch.zip"
+        $archive = Join-Path $tempDir $asset
+        $extractDir = Join-Path $tempDir $binary
+
+        Write-Host "Downloading $asset..."
+        Invoke-WebRequest -Uri "$baseUrl/$asset" -OutFile $archive
+        Expand-Archive -Path $archive -DestinationPath $extractDir -Force
+
+        $source = Join-Path $extractDir "$binary.exe"
+        if (-not (Test-Path $source)) {
+            throw "Release asset $asset did not contain $binary.exe."
+        }
+
+        Copy-Item -Path $source -Destination (Join-Path $InstallDir "$binary.exe") -Force
+    }
+
+    if (-not $NoPathUpdate) {
+        $currentPath = [Environment]::GetEnvironmentVariable("Path", "User")
+        $pathEntries = $currentPath -split ";" | Where-Object { $_ }
+        if ($pathEntries -notcontains $InstallDir) {
+            $newPath = if ([string]::IsNullOrWhiteSpace($currentPath)) { $InstallDir } else { "$currentPath;$InstallDir" }
+            [Environment]::SetEnvironmentVariable("Path", $newPath, "User")
+            Write-Host "Added $InstallDir to your user PATH. Restart your shell before running juleson."
+        }
+    }
+
+    Write-Host "Installed juleson and jules-mcp to $InstallDir"
+} finally {
+    Remove-Item -Recurse -Force $tempDir -ErrorAction SilentlyContinue
+}

--- a/scripts/install.sh
+++ b/scripts/install.sh
@@ -1,0 +1,150 @@
+#!/usr/bin/env bash
+set -euo pipefail
+
+repo="SamyRai/juleson"
+version="latest"
+install_dir="${INSTALL_DIR:-/usr/local/bin}"
+
+usage() {
+	cat <<'EOF'
+Install the latest Juleson release binaries.
+
+Usage:
+  install.sh [--version <tag|latest>] [--install-dir <path>] [--repo <owner/repo>]
+
+Environment:
+  INSTALL_DIR  Destination directory. Defaults to /usr/local/bin.
+
+Examples:
+  curl -L https://github.com/SamyRai/juleson/releases/latest/download/install.sh | bash
+  INSTALL_DIR="$HOME/.local/bin" bash install.sh
+  bash install.sh --version v1.0.0 --install-dir "$HOME/bin"
+EOF
+}
+
+while [ "$#" -gt 0 ]; do
+	case "$1" in
+		--version)
+			version="${2:-}"
+			shift 2
+			;;
+		--install-dir)
+			install_dir="${2:-}"
+			shift 2
+			;;
+		--repo)
+			repo="${2:-}"
+			shift 2
+			;;
+		-h|--help)
+			usage
+			exit 0
+			;;
+		*)
+			echo "Unknown argument: $1" >&2
+			usage >&2
+			exit 2
+			;;
+	esac
+done
+
+if [ -z "$version" ] || [ -z "$install_dir" ] || [ -z "$repo" ]; then
+	echo "version, install directory, and repository must be non-empty" >&2
+	exit 2
+fi
+
+require_cmd() {
+	if ! command -v "$1" >/dev/null 2>&1; then
+		echo "Required command not found: $1" >&2
+		exit 1
+	fi
+}
+
+require_cmd curl
+require_cmd tar
+require_cmd install
+require_cmd mktemp
+
+case "$(uname -s)" in
+	Linux)
+		os="linux"
+		;;
+	Darwin)
+		os="darwin"
+		;;
+	*)
+		echo "Unsupported operating system: $(uname -s)" >&2
+		exit 1
+		;;
+esac
+
+case "$(uname -m)" in
+	x86_64|amd64)
+		arch="amd64"
+		;;
+	arm64|aarch64)
+		arch="arm64"
+		;;
+	*)
+		echo "Unsupported architecture: $(uname -m)" >&2
+		exit 1
+		;;
+esac
+
+if [ "$version" = "latest" ]; then
+	base_url="https://github.com/${repo}/releases/latest/download"
+else
+	base_url="https://github.com/${repo}/releases/download/${version}"
+fi
+
+tmp_dir="$(mktemp -d)"
+cleanup() {
+	rm -rf "$tmp_dir"
+}
+trap cleanup EXIT
+
+download_and_extract() {
+	binary="$1"
+	asset="${binary}-${os}-${arch}.tar.gz"
+	archive="${tmp_dir}/${asset}"
+
+	echo "Downloading ${asset}..."
+	curl -fsSL "${base_url}/${asset}" -o "$archive"
+
+	mkdir -p "${tmp_dir}/${binary}"
+	tar -xzf "$archive" -C "${tmp_dir}/${binary}"
+	if [ ! -f "${tmp_dir}/${binary}/${binary}" ]; then
+		echo "Release asset ${asset} did not contain ${binary}" >&2
+		exit 1
+	fi
+}
+
+download_and_extract "juleson"
+download_and_extract "jules-mcp"
+
+mkdir -p "$install_dir" 2>/dev/null || true
+
+install_binary() {
+	binary="$1"
+	source="${tmp_dir}/${binary}/${binary}"
+	target="${install_dir}/${binary}"
+
+	if [ -w "$install_dir" ]; then
+		install -m 0755 "$source" "$target"
+	elif command -v sudo >/dev/null 2>&1; then
+		sudo mkdir -p "$install_dir"
+		sudo install -m 0755 "$source" "$target"
+	else
+		echo "Cannot write to ${install_dir}; rerun with a writable INSTALL_DIR or install sudo" >&2
+		exit 1
+	fi
+}
+
+install_binary "juleson"
+install_binary "jules-mcp"
+
+echo "Installed juleson and jules-mcp to ${install_dir}"
+case ":$PATH:" in
+	*":${install_dir}:"*) ;;
+	*) echo "Add ${install_dir} to PATH before running juleson." ;;
+esac

--- a/scripts/install.sh
+++ b/scripts/install.sh
@@ -91,7 +91,9 @@ case "$(uname -m)" in
 		;;
 esac
 
-if [ "$version" = "latest" ]; then
+if [ -n "${JULESON_INSTALL_BASE_URL:-}" ]; then
+	base_url="${JULESON_INSTALL_BASE_URL%/}"
+elif [ "$version" = "latest" ]; then
 	base_url="https://github.com/${repo}/releases/latest/download"
 else
 	base_url="https://github.com/${repo}/releases/download/${version}"

--- a/scripts/install.sh
+++ b/scripts/install.sh
@@ -52,6 +52,9 @@ if [ -z "$version" ] || [ -z "$install_dir" ] || [ -z "$repo" ]; then
 	echo "version, install directory, and repository must be non-empty" >&2
 	exit 2
 fi
+if [ "$install_dir" != "/" ]; then
+	install_dir="${install_dir%/}"
+fi
 
 require_cmd() {
 	if ! command -v "$1" >/dev/null 2>&1; then

--- a/scripts/install_test.go
+++ b/scripts/install_test.go
@@ -112,7 +112,7 @@ func TestReleaseWorkflowPublishesInstallAssets(t *testing.T) {
 	if err != nil {
 		t.Fatal(err)
 	}
-	text := string(workflow)
+	text := strings.ReplaceAll(string(workflow), "\r\n", "\n")
 
 	for _, want := range []string{
 		"cp scripts/install.sh dist/install.sh",

--- a/scripts/install_test.go
+++ b/scripts/install_test.go
@@ -1,0 +1,196 @@
+package main
+
+import (
+	"archive/tar"
+	"bytes"
+	"compress/gzip"
+	"fmt"
+	"io"
+	"net/http"
+	"net/http/httptest"
+	"os"
+	"os/exec"
+	"path/filepath"
+	"runtime"
+	"strings"
+	"testing"
+)
+
+func TestInstallShellHelpAndSyntax(t *testing.T) {
+	if runtime.GOOS == "windows" {
+		t.Skip("shell installer is for Linux/macOS")
+	}
+
+	syntax := exec.Command("bash", "-n", "install.sh")
+	if output, err := syntax.CombinedOutput(); err != nil {
+		t.Fatalf("install.sh syntax check failed: %v\n%s", err, output)
+	}
+
+	help := exec.Command("bash", "install.sh", "--help")
+	output, err := help.CombinedOutput()
+	if err != nil {
+		t.Fatalf("install.sh --help failed: %v\n%s", err, output)
+	}
+	for _, want := range []string{"--version", "--install-dir", "INSTALL_DIR"} {
+		if !strings.Contains(string(output), want) {
+			t.Fatalf("install.sh help missing %q:\n%s", want, output)
+		}
+	}
+}
+
+func TestInstallShellInstallsBothBinariesFromLocalRelease(t *testing.T) {
+	if testing.Short() {
+		t.Skip("skipping installer E2E test in short mode")
+	}
+	if runtime.GOOS == "windows" {
+		t.Skip("shell installer is for Linux/macOS")
+	}
+
+	goos, goarch, ok := installerPlatform()
+	if !ok {
+		t.Skipf("installer does not support %s/%s", runtime.GOOS, runtime.GOARCH)
+	}
+
+	releaseDir := t.TempDir()
+	for _, binary := range []string{"juleson", "jules-mcp"} {
+		asset := fmt.Sprintf("%s-%s-%s.tar.gz", binary, goos, goarch)
+		if err := writeTarGz(filepath.Join(releaseDir, asset), binary, []byte("#!/bin/sh\necho "+binary+"\n")); err != nil {
+			t.Fatal(err)
+		}
+	}
+
+	server := httptest.NewServer(http.FileServer(http.Dir(releaseDir)))
+	t.Cleanup(server.Close)
+
+	installDir := filepath.Join(t.TempDir(), "bin")
+	cmd := exec.Command("bash", "install.sh", "--version", "v0.0.0-test", "--install-dir", installDir)
+	cmd.Env = append(os.Environ(), "JULESON_INSTALL_BASE_URL="+server.URL)
+	output, err := cmd.CombinedOutput()
+	if err != nil {
+		t.Fatalf("install.sh failed: %v\n%s", err, output)
+	}
+
+	for _, binary := range []string{"juleson", "jules-mcp"} {
+		path := filepath.Join(installDir, binary)
+		info, err := os.Stat(path)
+		if err != nil {
+			t.Fatalf("expected installed binary %s: %v\ninstaller output:\n%s", path, err, output)
+		}
+		if info.Mode()&0111 == 0 {
+			t.Fatalf("expected %s to be executable, mode %s", path, info.Mode())
+		}
+		contents, err := os.ReadFile(path)
+		if err != nil {
+			t.Fatal(err)
+		}
+		if !bytes.Contains(contents, []byte("echo "+binary)) {
+			t.Fatalf("installed %s has unexpected contents: %q", binary, contents)
+		}
+	}
+}
+
+func TestPowerShellInstallerHelpWhenAvailable(t *testing.T) {
+	pwsh, err := exec.LookPath("pwsh")
+	if err != nil {
+		t.Skip("pwsh not available")
+	}
+
+	cmd := exec.Command(pwsh, "-NoProfile", "-File", "install.ps1", "-Help")
+	output, err := cmd.CombinedOutput()
+	if err != nil {
+		t.Fatalf("install.ps1 -Help failed: %v\n%s", err, output)
+	}
+	for _, want := range []string{"-Version", "-InstallDir", "-BaseUrl"} {
+		if !strings.Contains(string(output), want) {
+			t.Fatalf("install.ps1 help missing %q:\n%s", want, output)
+		}
+	}
+}
+
+func TestReleaseWorkflowPublishesInstallAssets(t *testing.T) {
+	workflow, err := os.ReadFile(filepath.Join("..", ".github", "workflows", "release.yml"))
+	if err != nil {
+		t.Fatal(err)
+	}
+	text := string(workflow)
+
+	for _, want := range []string{
+		"cp scripts/install.sh dist/install.sh",
+		"cp scripts/install.ps1 dist/install.ps1",
+		"juleson-${OS}-${ARCH}.tar.gz",
+		"jules-mcp-${OS}-${ARCH}.tar.gz",
+		"juleson-${OS}-${ARCH}.zip",
+		"jules-mcp-${OS}-${ARCH}.zip",
+		"github.com/SamyRai/juleson@${{ needs.validate.outputs.version }}",
+		"if: github.event_name == 'workflow_dispatch'",
+	} {
+		if !strings.Contains(text, want) {
+			t.Fatalf("release workflow missing %q", want)
+		}
+	}
+}
+
+func TestInstallationDocsReferenceInstallableAssets(t *testing.T) {
+	files := []string{
+		filepath.Join("..", "README.md"),
+		filepath.Join("..", "docs", "INSTALLATION_GUIDE.md"),
+		filepath.Join("README.md"),
+	}
+	for _, file := range files {
+		contents, err := os.ReadFile(file)
+		if err != nil {
+			t.Fatal(err)
+		}
+		text := string(contents)
+		if strings.Contains(text, "cmd/juleson-mcp") {
+			t.Fatalf("%s still references removed cmd/juleson-mcp path", file)
+		}
+		if strings.Contains(text, "Go 1.23+") || strings.Contains(text, "Go 1.24+") || strings.Contains(text, "Go 1.23 or higher") || strings.Contains(text, "Go 1.24 or higher") {
+			t.Fatalf("%s still references an outdated Go prerequisite", file)
+		}
+	}
+}
+
+func installerPlatform() (string, string, bool) {
+	var goos string
+	switch runtime.GOOS {
+	case "linux", "darwin":
+		goos = runtime.GOOS
+	default:
+		return "", "", false
+	}
+
+	var goarch string
+	switch runtime.GOARCH {
+	case "amd64", "arm64":
+		goarch = runtime.GOARCH
+	default:
+		return "", "", false
+	}
+	return goos, goarch, true
+}
+
+func writeTarGz(path, name string, contents []byte) error {
+	file, err := os.Create(path)
+	if err != nil {
+		return err
+	}
+	defer file.Close()
+
+	gzipWriter := gzip.NewWriter(file)
+	defer gzipWriter.Close()
+
+	tarWriter := tar.NewWriter(gzipWriter)
+	defer tarWriter.Close()
+
+	header := &tar.Header{
+		Name: name,
+		Mode: 0755,
+		Size: int64(len(contents)),
+	}
+	if err := tarWriter.WriteHeader(header); err != nil {
+		return err
+	}
+	_, err = io.Copy(tarWriter, bytes.NewReader(contents))
+	return err
+}

--- a/scripts/install_test.go
+++ b/scripts/install_test.go
@@ -121,6 +121,7 @@ func TestReleaseWorkflowPublishesInstallAssets(t *testing.T) {
 		"jules-mcp-${OS}-${ARCH}.tar.gz",
 		"juleson-${OS}-${ARCH}.zip",
 		"jules-mcp-${OS}-${ARCH}.zip",
+		"- goos: windows\n            goarch: arm64",
 		"github.com/SamyRai/juleson@${{ needs.validate.outputs.version }}",
 		"if: github.event_name == 'workflow_dispatch'",
 	} {

--- a/scripts/install_test.go
+++ b/scripts/install_test.go
@@ -76,7 +76,7 @@ func TestInstallShellInstallsBothBinariesFromLocalRelease(t *testing.T) {
 		if err != nil {
 			t.Fatalf("expected installed binary %s: %v\ninstaller output:\n%s", path, err, output)
 		}
-		if info.Mode()&0111 == 0 {
+		if info.Mode()&0o111 == 0 {
 			t.Fatalf("expected %s to be executable, mode %s", path, info.Mode())
 		}
 		contents, err := os.ReadFile(path)
@@ -135,7 +135,7 @@ func TestInstallationDocsReferenceInstallableAssets(t *testing.T) {
 	files := []string{
 		filepath.Join("..", "README.md"),
 		filepath.Join("..", "docs", "INSTALLATION_GUIDE.md"),
-		filepath.Join("README.md"),
+		"README.md",
 	}
 	for _, file := range files {
 		contents, err := os.ReadFile(file)
@@ -152,8 +152,7 @@ func TestInstallationDocsReferenceInstallableAssets(t *testing.T) {
 	}
 }
 
-func installerPlatform() (string, string, bool) {
-	var goos string
+func installerPlatform() (goos, goarch string, ok bool) {
 	switch runtime.GOOS {
 	case "linux", "darwin":
 		goos = runtime.GOOS
@@ -161,7 +160,6 @@ func installerPlatform() (string, string, bool) {
 		return "", "", false
 	}
 
-	var goarch string
 	switch runtime.GOARCH {
 	case "amd64", "arm64":
 		goarch = runtime.GOARCH
@@ -186,7 +184,7 @@ func writeTarGz(path, name string, contents []byte) error {
 
 	header := &tar.Header{
 		Name: name,
-		Mode: 0755,
+		Mode: 0o755,
 		Size: int64(len(contents)),
 	}
 	if err := tarWriter.WriteHeader(header); err != nil {

